R/save.R

```diff
@@ -189,12 +196,48 @@ parse_dpi <- function(dpi, call = caller_env()) {
   }
 }
 
-plot_dim <- function(dim = c(NA, NA), scale = 1, units = "in",
-                     limitsize = TRUE, dpi = 300, call = caller_env()) {
+plot_dim <- function(width = NA, height = NA, scale = 1, units = "in",
+                     limitsize = TRUE, dpi = 300, plot = NULL, call = caller_env()) {
   units <- arg_match0(units, c("in", "cm", "mm", "px"))
-  to_inches <- function(x) x / c(`in` = 1, cm = 2.54, mm = 2.54 * 10, px = dpi)[units]
+  to_inches   <- function(x) x / c(`in` = 1, cm = 2.54, mm = 2.54 * 10, px = dpi)[units]
   from_inches <- function(x) x * c(`in` = 1, cm = 2.54, mm = 2.54 * 10, px = dpi)[units]
 
+  if (is.derived(width) || is.derived(height)) {
+    # To size from plot if width or height are derived
+    # TODO: use gtable::as.gtable when implemented
+    if (is.ggplot(plot)) {
+      plot <- ggplotGrob(plot)
+    }
+    if (!inherits(plot, "gtable")) {
+      cli::cli_abort(
+        "Cannot derive size of plot when {.arg plot} is \\
+        {.obj_type_friendly {plot}}.",
+        call = call
+      )
+    }
+    width  <- if (is.derived(width))  gtable_width(plot)  else width
+    height <- if (is.derived(height)) gtable_height(plot) else height
+  }
+
+  if (is.unit(width)) {
+    if (has_null_unit(width)) {
```

Comment:
The assumption here is that if the plot dimensions contain `"null"` units, the dimension is relative. If these are absent, the plot is assumed to have absolute dimensions. 

It is possible for facet or axis guide extensions to break this assumption (as did `facet_null()`).